### PR TITLE
POC: S3 support for OpenHouse 

### PR DESCRIPTION
**Pull request by HotSushi**
## Summary
This is a draft PR to showcase s3 support for OpenHouse. Please see the code comments for what needs to be done thoroughly.

Oh! and watch this video as well:

https://github.com/linkedin/openhouse/assets/6441597/dacf0844-5457-4467-85eb-113eccbb9567

### CODE REVIEW

File: cluster/storage/build.gradle [hunk 1]
```diff
@@ -7,4 +7,6 @@ plugins {
 dependencies {
   implementation project(':cluster:configs')
   implementation 'org.springframework.boot:spring-boot-autoconfigure:' + spring_web_version
```
Comment (HotSushi) [inline cluster/storage/build.gradle:9]: need to move these dependencies to "s3-conventions.gradle"

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  @Autowired private StorageProperties storageProperties;
+
+  private S3Client s3;
+
+  private static final StorageType.Type S3_TYPE = StorageType.S3;
+
+  @PostConstruct
+  public synchronized void init() {
+    validateProperties();
```
Comment (HotSushi) [inline cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java:32]: validateProperties is a common logic, need to move it to: `BaseStorageClient`

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/fileio/FileIOConfig.java [hunk 3]
```diff
@@ -62,4 +70,67 @@ FileIO provideLocalFileIO() {
...
+    } catch (IllegalArgumentException e) {
+      // If the S3 storage type is not configured, return null
+      // Spring doesn't define the bean if the return value is null
+      log.debug("S3 storage type is not configured", e);
+      return null;
+    }
+  }
+
+  @AllArgsConstructor
+  public static class DelegatingInputFile implements InputFile {
```
Comment (HotSushi) [inline iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/fileio/FileIOConfig.java:105]: seem hacky, is there a better way to solve this? should OH get rid of non-schemed paths completely?

File: SETUP.md [hunk 4]
```diff
@@ -231,15 +231,21 @@ docker exec -it local.spark-master /bin/bash
...
   --jars openhouse-spark-runtime_2.12-*-all.jar  \
   --conf spark.sql.extensions=org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions,com.linkedin.openhouse.spark.extensions.OpenhouseSparkSessionExtensions   \
   --conf spark.sql.catalog.openhouse=org.apache.iceberg.spark.SparkCatalog   \
   --conf spark.sql.catalog.openhouse.catalog-impl=com.linkedin.openhouse.spark.OpenHouseCatalog     \
   --conf spark.sql.catalog.openhouse.metrics-reporter-impl=com.linkedin.openhouse.javaclient.OpenHouseMetricsReporter    \
   --conf spark.sql.catalog.openhouse.uri=http://openhouse-tables:8080   \
   --conf spark.sql.catalog.openhouse.auth-token=$(cat /var/config/$(whoami).token) \
-  --conf spark.sql.catalog.openhouse.cluster=LocalHadoopCluster
+  --conf spark.sql.catalog.openhouse.cluster=LocalFSCluster  \
+  --conf spark.sql.catalog.openhouse.com.linkedin.openhouse.relocated.io-impl=org.apache.iceberg.aws.s3.S3FileIO  \
```
Comment (HotSushi) [inline SETUP.md:244]: this is a bug in relocations in java-runtime should be `spark.sql.catalog.openhouse.io-impl` instead.

**Review by HotSushi — COMMENTED** [review]
cc: @jainlavina , @sumedhsakdeo , @autumnust, @ctrezzo

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/fileio/FileIOConfig.java (hunk 3 above)
Comment (sumedhsakdeo) [reply iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/fileio/FileIOConfig.java:105]: OpenHouseS3FileIO with delegation to S3FileIO makes sense to me even for implementing the future token or signed url based authentication.

**Comment by HotSushi** [comment]
Code is already checked in and running well in production. Getting started guide is published here: https://www.openhousedb.org/docs/getting_started

**Pull request closed without merging**